kimchi/src/circuits/polynomials/foreign_field_add/witness.rs

```diff
@@ -127,74 +160,134 @@ pub fn create<F: PrimeField>(
         // We only obtain the 3 lower limbs of right because the range check takes only 264 bits now
         let right_3_limb = ForeignElement::new([right[LO], right[MI], right[HI]]);
-        range_check::witness::extend(&mut witness, right_3_limb);
-        range_check::witness::extend(&mut witness, output.clone());
+        // Create multi-range-check witness for right_input (left_input was done in previous iteration) and output
+        if range_checks {
+            range_check::witness::extend(&mut witness, right_3_limb);
+            range_check::witness::extend(&mut witness, add_values.output.clone());
+        }
 
-        add_values.append(&mut vec![(sign, overflow, carry_lo, carry_mi)]);
-        left = output; // output
+        values.push(add_values.clone());
+        left = add_values.output.clone(); // output
     }
 
     // Compute values for final bound check, needs a 4 limb right input
     let right = ForeignElement::<F, 4>::from_biguint(BigUint::from(TWO_TO_LIMB).pow(3));
 
-    let (bound, sign, overflow, bound_carry_lo, bound_carry_mi) =
-        compute_subadd_values(&left, &right, FFOps::Add, &foreign_modulus);
+    let bound_values = compute_ffadd_values(&left, &right, FFOps::Add, &foreign_modulus);
     // Make sure they have the right value
-    assert_eq!(sign, F::one());
-    assert_eq!(overflow, F::one());
+    assert_eq!(bound_values.sign, F::one());
+    assert_eq!(bound_values.ovf, F::one());
 
     // Final RangeCheck for bound
-    range_check::witness::extend(&mut witness, bound);
-    let mut offset = witness[LO].len(); // number of witness rows of the gadget before the first row of the addition gate
+    if range_checks {
+        range_check::witness::extend(&mut witness, bound_values.output.clone());
+    }
+    let mut offset = if range_checks {
+        witness[LO].len() // number of witness rows of the gadget before the first row of the addition gate
+    } else {
+        0
+    };
 
     // Include FFAdds gates for operations and final bound check
 
-    for (i, value) in add_values.iter().enumerate() {
+    for (i, value) in values.iter().enumerate() {
         // Create foreign field addition row
         for w in &mut witness {
             w.extend(std::iter::repeat(F::zero()).take(1));
         }
 
-        let (sign, overflow, carry_lo, carry_mi) = *value;
-
         // ForeignFieldAdd row and Zero row
-        init_foreign_field_add_rows(
-            &mut witness,
-            offset,
-            i,
-            sign,
-            overflow,
-            [carry_lo, carry_mi],
-        );
+        if range_checks {
+            init_ff_add_rows_rc(&mut witness, offset, i, value.sign, value.ovf, value.carry);
+        } else {
+            let right = ForeignElement::new([value.right[LO], value.right[MI], value.right[HI]]);
+            init_ff_add_rows(
+                &mut witness,
+                offset,
+                value.left.clone(),
+                right,
```

Comment:
Same as above,  now you can just do `value.right.limbs` and completely avoid the extra `let right = ForeignElement::new([value.right[LO], value.right[MI], value.right[HI]]);` line.

Reply:
do you mean if I change the input type of the init_ffadd_row function?

Reply:
yup